cylc/flow/platform_lookup.py

```diff
@@ -0,0 +1,101 @@
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Tests for the platform lookup.
+
+import re
+from cylc.flow.exceptions import PlatformLookupError
+
+
+def forward_lookup(platforms, task_platform):
+    """
+    Find out which job platform to use given a list of possible platforms and
+    a task platform string.
+
+    Verifies selected platform is present in global.rc file and returns it,
+    raises error if platfrom is not in global.rc or returns 'localhost' if
+    no platform is initally selected.
+
+    Args:
+        task_platform (str):
+            platform item from config [runtime][TASK][platform]
+        platforms (dictionary):
+            list of possible platforms defined by global.rc
+
+    Returns:
+        platform (str):
+            string representing a platform from the global config.
+
+    Example:
+    Example Input:
+    >>> platforms = {
+    ...     'suite server platform': None,
+    ...     'desktop[0-9][0-9]|laptop[0-9][0-9]': None,
+    ...     'sugar': {
+    ...         'login hosts': 'localhost',
+    ...         'batch system': 'slurm'
+    ...     },
+    ...     'hpc': {
+    ...         'login hosts': ['hpc1', 'hpc2'],
+    ...         'batch system': 'pbs'
+    ...     },
+    ...     'hpc1-bg': {
+    ...         'login hosts': 'hpc1',
+    ...         'batch system': 'background'
+    ...     },
+    ...     'hpc2-bg': {
+    ...         'login hosts': 'hpc2',
+    ...         'batch system': 'background'
+    ...     }
+    ... }
+    >>> task_platform = 'desktop22'
+    >>> forward_lookup(platforms, task_platform)
+    'desktop22'
+    """
+    if task_platform is None:
+        return 'localhost'
+    platforms = list(platforms.keys())
+    reversed_platforms = platforms[::-1]
```

Comment:
I've learned something today, have never seen that syntax before.

This one-liner might clarify things a little though:

```python
for platform in reversed(list(platforms))):
```

Reply:
Me too (not seen that!). As standard Python syntax, it's probably something we should start using (even though Oliver's more explicit syntax is clearer to the uninitiated).

Reply:
Not just me then!

Reply:
Brainwashed by data science slicing!
I'm outnumbered and will change to the more readable reversed :)